### PR TITLE
feat: add empty dir into wait container

### PR DESCRIPTION
**Pull request by haoxins**
Fixes https://github.com/argoproj/argo-workflows/issues/8376

### CODE REVIEW

File: workflow/controller/workflowpod.go [hunk 1]
```diff
@@ -37,6 +37,16 @@ var (
...
+	volumeTmpDirArgo = apiv1.Volume{
+		Name: "tmp-dir-argo",
+		VolumeSource: apiv1.VolumeSource{
+			EmptyDir: &apiv1.EmptyDirVolumeSource{},
+		},
+	}
+	volumeMountTmpDirArgo = apiv1.VolumeMount{
+		Name:      volumeTmpDirArgo.Name,
+		MountPath: common.TmpDirArgoPath,
+	}
```
Comment (rohankmr414) [inline workflow/controller/workflowpod.go:49]: I think here the root directory is being overwritten by `/tmp`, that is why the tests are failing. Perhaps you should try using `subPaths` 
see https://kubernetes.io/docs/concepts/storage/volumes/#using-subpath

Comment (haoxins) [reply workflow/controller/workflowpod.go:49]: thanks

File: workflow/common/common.go [hunk 2]
```diff
@@ -212,6 +212,9 @@ const (
 	// VarRunArgoPath is the standard path for the shared volume
 	VarRunArgoPath = "/var/run/argo"
 
+	// TmpDirArgoPath is the standard path for the tmp volume
```
Comment (alexec) [inline workflow/common/common.go:215]: I’m not sure we need a variable for this. It cannot be changed.

File: workflow/controller/workflowpod.go [hunk 3]
```diff
@@ -37,6 +37,17 @@ var (
 		Name:      volumeVarArgo.Name,
 		MountPath: common.VarRunArgoPath,
 	}
+	volumeTmpDirArgo = apiv1.Volume{
```
Comment (alexec) [inline workflow/controller/workflowpod.go:40]: I think you can loose the “Argo” suffix.

**Comment by haoxins** [comment]
@alexec @rohankmr414 
I'm not sure whether it's a good idea to mount the `tmp-dir` into all the containers (just for e2e testing pass)
And will be appreciative if you guys can help to take a look at the failed e2e tests. thanks~

**Comment by alexec** [comment]
Good question. I did ponder this. Options.

If we mount /tmp on every container, it makes it easy for all containers to run with ready-only root FS. Which is good for security. But, what if they don’t want a /tmp directory? Or their /tmp is not empty for some reason (can’t imagine why). 

If they want a /tmp directory, they can work-around anyway. 

We should not share the same /tmp volume. So it should only be mounted on the container named `wait`. There are other examples.

**Comment by alexec** [comment]
The failing tests are because the resource template container is named `main` not `wait`. 

Instead of checking the container name, check the image. If it is the same as `wfc.executorImage()` than add the volume.

**Comment by alexec** [comment]
```
⚠ k8s-resource-tmpl-with-pod-2hq42   Workflow 2s      Error (exit code 1): open /tmp/manifest.yaml: no such file or directory
[302](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:302)
 └ ⚠ k8s-resource-tmpl-with-pod-2hq42 Pod      1s      Error (exit code 1): open /tmp/manifest.yaml: no such file or directory
[303](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:303)

[304](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:304)
Condition "to be done" met after 2s
[305](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:305)
Checking expectation k8s-resource-tmpl-with-pod-2hq42
[306](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:306)
k8s-resource-tmpl-with-pod-2hq42 : Error Error (exit code 1): open /tmp/manifest.yaml: no such file or directory
[307](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:307)
    resource_template_test.go:95: 
[308](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:308)
        	Error Trace:	resource_template_test.go:95
[309](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:309)
        	            				then.go:67
[310](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:310)
        	            				then.go:42
[311](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:311)
        	            				resource_template_test.go:94
[312](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:312)
        	Error:      	Not equal: 
[313](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:313)
        	            	expected: "Succeeded"
[314](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:314)
        	            	actual  : "Error"
[315](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:315)
        	            	
[316](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:316)
        	            	Diff:
[317](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:317)
        	            	--- Expected
[318](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:318)
        	            	+++ Actual
[319](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:319)
        	            	@@ -1,2 +1,2 @@
[320](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:320)
        	            	-(v1alpha1.WorkflowPhase) (len=9) "Succeeded"
[321](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:321)
        	            	+(v1alpha1.WorkflowPhase) (len=5) "Error"
[322](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:322)
        	            	 
[323](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:323)
        	Test:       	TestResourceTemplateSuite/TestResourceTemplateWithPod
[324](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:324)
=== FAIL: ResourceTemplateSuite/TestResourceTemplateWithPod
```

You have test failures relating to resource template.

This line looks wrong

```golang
		MountPath: "/tmp/" + strconv.Itoa(i),
```

Should it not be?

```golang
		MountPath: "/tmp",
```

**Comment by haoxins** [comment]
> ```
> ⚠ k8s-resource-tmpl-with-pod-2hq42   Workflow 2s      Error (exit code 1): open /tmp/manifest.yaml: no such file or directory
> [302](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:302)
>  └ ⚠ k8s-resource-tmpl-with-pod-2hq42 Pod      1s      Error (exit code 1): open /tmp/manifest.yaml: no such file or directory
> [303](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:303)
> 
> [304](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:304)
> Condition "to be done" met after 2s
> [305](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:305)
> Checking expectation k8s-resource-tmpl-with-pod-2hq42
> [306](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:306)
> k8s-resource-tmpl-with-pod-2hq42 : Error Error (exit code 1): open /tmp/manifest.yaml: no such file or directory
> [307](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:307)
>     resource_template_test.go:95: 
> [308](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:308)
>         	Error Trace:	resource_template_test.go:95
> [309](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:309)
>         	            				then.go:67
> [310](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:310)
>         	            				then.go:42
> [311](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:311)
>         	            				resource_template_test.go:94
> [312](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:312)
>         	Error:      	Not equal: 
> [313](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:313)
>         	            	expected: "Succeeded"
> [314](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:314)
>         	            	actual  : "Error"
> [315](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:315)
>         	            	
> [316](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:316)
>         	            	Diff:
> [317](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:317)
>         	            	--- Expected
> [318](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:318)
>         	            	+++ Actual
> [319](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:319)
>         	            	@@ -1,2 +1,2 @@
> [320](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:320)
>         	            	-(v1alpha1.WorkflowPhase) (len=9) "Succeeded"
> [321](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:321)
>         	            	+(v1alpha1.WorkflowPhase) (len=5) "Error"
> [322](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:322)
>         	            	 
> [323](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:323)
>         	Test:       	TestResourceTemplateSuite/TestResourceTemplateWithPod
> [324](https://github.com/argoproj/argo-workflows/runs/6066426195?check_suite_focus=true#step:16:324)
> === FAIL: ResourceTemplateSuite/TestResourceTemplateWithPod
> ```
> 
> You have test failures relating to resource template.
> 
> This line looks wrong
> 
> ```go
> 		MountPath: "/tmp/" + strconv.Itoa(i),
> ```
> 
> Should it not be?
> 
> ```go
> 		MountPath: "/tmp",
> ```

MountPath: "/tmp" will get the same errors :) a bit strange

**Review by alexec — COMMENTED** [review]
maybe try without Subpath?

**Comment by haoxins** [comment]
> maybe try without Subpath?

the same errs :)

**Comment by alexec** [comment]
Have you started this up locally?

File: Dockerfile [hunk 4]
```diff
@@ -98,6 +98,7 @@ COPY --from=argoexec-build /usr/local/bin/kubectl /bin/
 COPY --from=argoexec-build /usr/local/bin/jq /bin/
```
Comment (haoxins) [inline Dockerfile:98]: @alexec I'm not sure why but the image is the root cause.

**Comment by haoxins** [comment]
https://www.reddit.com/r/docker/comments/8y2zyx/how_to_add_a_empty_directory_to_a_scratch_image/hneaaqe/?utm_source=share&utm_medium=web2x&context=3

File: Dockerfile (hunk 4 above)
Comment (haoxins) [reply Dockerfile:98]: https://www.reddit.com/r/docker/comments/8y2zyx/how_to_add_a_empty_directory_to_a_scratch_image/hneaaqe/

**Review by alexec — APPROVED** [review]

**Comment by huiwq1990** [comment]
@haoxins Why we use VolumeMounts subpath mount dir, could we use VolumeMounts mountpath? As we use k8s1.19, which has a subpath bug.

**Pull request merged**